File: sdks/typescript/tests/integration.test.ts
import { describe, it, expect, beforeAll } from "vitest";
import { createClient, type KraalzibarClient } from "../src/index.js";

const TARGET = process.env.KRAALZIBAR_REST_TARGET ?? "http://localhost:8080";
const isIntegration = process.env.INTEGRATION === "true";

describe.skipIf(!isIntegration)("integration", () => {
  let client: KraalzibarClient;

  beforeAll(() => {
    client = createClient({ target: TARGET });
  });

  const writeSchema = async (schema: string) => {
    await client.writeSchema(schema, { force: true });
  };

  const touchRelationship = async (
    resourceType: string,
    resourceId: string,
    relation: string,
    subjectType: string,
    subjectId: string,
  ) => {
    await client.writeRelationships({
      updates: [
        {
          operation: "touch",
          resource_type: resourceType,
          resource_id: resourceId,
          relation,
          subject_type: subjectType,
          subject_id: subjectId,
        },
      ],
    });
  };

  const basicSchema = `definition user {}

definition document {
  relation viewer: user
  permission can_view = viewer
}`;

  it("check permission granted", async () => {
    await writeSchema(basicSchema);
    await touchRelationship("document", "readme", "viewer", "user", "alice");

    const result = await client.checkPermission({
      resource_type: "document",
      resource_id: "readme",
      permission: "can_view",
      subject_type: "user",
      subject_id: "alice",
      consistency: { type: "full" },
    });

    expect(result.allowed).toBe(true);
  });

  it("check permission denied", async () => {
    await writeSchema(basicSchema);
    await touchRelationship("document", "readme", "viewer", "user", "alice");

    const result = await client.checkPermission({
      resource_type: "document",
      resource_id: "readme",
      permission: "can_view",
      subject_type: "user",
      subject_id: "bob",
      consistency: { type: "full" },
    });

    expect(result.allowed).toBe(false);
  });

  it("write and read relationships", async () => {
    await writeSchema(basicSchema);
    await touchRelationship("document", "readme", "viewer", "user", "alice");
    await touchRelationship("document", "readme", "viewer", "user", "bob");

    const result = await client.readRelationships({
      filter: {
        resource_type: "document",
        resource_id: "readme",
        relation: "viewer",
      },
      consistency: { type: "full" },
    });

    expect(result.relationships.length).toBe(2);
  });

  it("lookup resources", async () => {
    await writeSchema(basicSchema);
    await touchRelationship("document", "doc1", "viewer", "user", "alice");
    await touchRelationship("document", "doc2", "viewer", "user", "alice");
    await touchRelationship("document", "doc3", "viewer", "user", "bob");

    const result = await client.lookupResources({
      resource_type: "document",
      permission: "can_view",
      subject_type: "user",
      subject_id: "alice",
      consistency: { type: "full" },
    });

    const ids = [...result.resource_ids].sort();
    expect(ids).toEqual(["doc1", "doc2"]);
  });

  it("lookup subjects", async () => {
    await writeSchema(`definition user {}

definition document {
  relation viewer: user
  relation owner: user
  permission can_view = viewer + owner
}`);
    await touchRelationship("document", "readme", "viewer", "user", "alice");
    await touchRelationship("document", "readme", "owner", "user", "bob");

    const result = await client.lookupSubjects({
      resource_type: "document",
      resource_id: "readme",
      permission: "can_view",
      subject_type: "user",
      consistency: { type: "full" },
    });

    const ids = result.subjects.map((s) => s.subject_id).sort();
    expect(ids).toEqual(["alice", "bob"]);
  });

  it("schema write and read", async () => {
    await writeSchema(basicSchema);

    const schema = await client.readSchema();

    expect(schema).toContain("definition user");
    expect(schema).toContain("definition document");
    expect(schema).toContain("relation viewer");
    expect(schema).toContain("permission can_view");
  });

  it("computed permission via union", async () => {
    await writeSchema(`definition user {}

definition document {
  relation viewer: user
  relation editor: user
  permission can_edit = editor
  permission can_view = viewer + can_edit
}`);
    await touchRelationship("document", "readme", "editor", "user", "alice");

    const result = await client.checkPermission({
      resource_type: "document",
      resource_id: "readme",
      permission: "can_view",
      subject_type: "user",
      subject_id: "alice",
      consistency: { type: "full" },
    });

    expect(result.allowed).toBe(true);
  });
});

File: sdks/go/integration_test.go
//go:build integration

package kraalzibar

import (
	"context"
	"os"
	"sort"
	"testing"
)

func serverTarget() string {
	if t := os.Getenv("KRAALZIBAR_TARGET"); t != "" {
		return t
	}
	return "localhost:50051"
}

func testClient(t *testing.T) *Client {
	t.Helper()
	c, err := NewClient(serverTarget(), WithInsecure())
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	t.Cleanup(func() { _ = c.Close() })
	return c
}

func setupSchema(t *testing.T, c *Client, schema string) {
	t.Helper()
	ctx := context.Background()
	_, err := c.WriteSchema(ctx, schema, true)
	if err != nil {
		t.Fatalf("WriteSchema: %v", err)
	}
}

func touchRelationship(t *testing.T, c *Client, objectType, objectID, relation, subjectType, subjectID string) {
	t.Helper()
	ctx := context.Background()
	_, err := c.WriteRelationships(ctx, WriteRelationshipsRequest{
		Updates: []RelationshipUpdate{{
			Operation: OperationTouch,
			Object:    ObjectRef{ObjectType: objectType, ObjectID: objectID},
			Relation:  relation,
			Subject:   DirectSubject(subjectType, subjectID),
		}},
	})
	if err != nil {
		t.Fatalf("WriteRelationships: %v", err)
	}
}

const basicSchema = `definition user {}

definition document {
  relation viewer: user
  permission can_view = viewer
}`

func TestIntegration_CheckPermissionGranted(t *testing.T) {
	c := testClient(t)
	setupSchema(t, c, basicSchema)
	touchRelationship(t, c, "document", "readme", "viewer", "user", "alice")

	resp, err := c.CheckPermission(context.Background(), CheckPermissionRequest{
		Resource:   ObjectRef{ObjectType: "document", ObjectID: "readme"},
		Permission: "can_view",
		Subject:    DirectSubject("user", "alice"),
	})
	if err != nil {
		t.Fatalf("CheckPermission: %v", err)
	}
	if !resp.Allowed {
		t.Error("expected permission to be granted")
	}
}

func TestIntegration_CheckPermissionDenied(t *testing.T) {
	c := testClient(t)
	setupSchema(t, c, basicSchema)
	touchRelationship(t, c, "document", "readme", "viewer", "user", "alice")

	resp, err := c.CheckPermission(context.Background(), CheckPermissionRequest{
		Resource:   ObjectRef{ObjectType: "document", ObjectID: "readme"},
		Permission: "can_view",
		Subject:    DirectSubject("user", "bob"),
	})
	if err != nil {
		t.Fatalf("CheckPermission: %v", err)
	}
	if resp.Allowed {
		t.Error("expected permission to be denied")
	}
}

func TestIntegration_WriteAndReadRelationships(t *testing.T) {
	c := testClient(t)
	setupSchema(t, c, basicSchema)
	touchRelationship(t, c, "document", "readme", "viewer", "user", "alice")
	touchRelationship(t, c, "document", "readme", "viewer", "user", "bob")

	rels, err := c.ReadRelationships(context.Background(), ReadRelationshipsRequest{
		ObjectType: "document",
		ObjectID:   "readme",
		Relation:   "viewer",
	})
	if err != nil {
		t.Fatalf("ReadRelationships: %v", err)
	}
	if len(rels) != 2 {
		t.Errorf("expected 2 relationships, got %d", len(rels))
	}
}

func TestIntegration_LookupResources(t *testing.T) {
	c := testClient(t)
	setupSchema(t, c, basicSchema)
	touchRelationship(t, c, "document", "doc1", "viewer", "user", "alice")
	touchRelationship(t, c, "document", "doc2", "viewer", "user", "alice")
	touchRelationship(t, c, "document", "doc3", "viewer", "user", "bob")

	ids, err := c.LookupResources(context.Background(), LookupResourcesRequest{
		ResourceType: "document",
		Permission:   "can_view",
		Subject:      DirectSubject("user", "alice"),
	})
	if err != nil {
		t.Fatalf("LookupResources: %v", err)
	}
	sort.Strings(ids)
	if len(ids) != 2 || ids[0] != "doc1" || ids[1] != "doc2" {
		t.Errorf("expected [doc1, doc2], got %v", ids)
	}
}

func TestIntegration_LookupSubjects(t *testing.T) {
	c := testClient(t)
	setupSchema(t, c, `definition user {}

definition document {
  relation viewer: user
  relation owner: user
  permission can_view = viewer + owner
}`)
	touchRelationship(t, c, "document", "readme", "viewer", "user", "alice")
	touchRelationship(t, c, "document", "readme", "owner", "user", "bob")

	subjects, err := c.LookupSubjects(context.Background(), LookupSubjectsRequest{
		Resource:    ObjectRef{ObjectType: "document", ObjectID: "readme"},
		Permission:  "can_view",
		SubjectType: "user",
	})
	if err != nil {
		t.Fatalf("LookupSubjects: %v", err)
	}
	ids := make([]string, len(subjects))
	for i, s := range subjects {
		ids[i] = s.SubjectID
	}
	sort.Strings(ids)
	if len(ids) != 2 || ids[0] != "alice" || ids[1] != "bob" {
		t.Errorf("expected [alice, bob], got %v", ids)
	}
}

func TestIntegration_SchemaWriteAndRead(t *testing.T) {
	c := testClient(t)
	setupSchema(t, c, basicSchema)

	schema, err := c.ReadSchema(context.Background())
	if err != nil {
		t.Fatalf("ReadSchema: %v", err)
	}
	for _, expected := range []string{"definition user", "definition document", "relation viewer", "permission can_view"} {
		if !containsStr(schema, expected) {
			t.Errorf("schema missing %q", expected)
		}
	}
}

func TestIntegration_ComputedPermissionViaUnion(t *testing.T) {
	c := testClient(t)
	setupSchema(t, c, `definition user {}

definition document {
  relation viewer: user
  relation editor: user
  permission can_edit = editor
  permission can_view = viewer + can_edit
}`)
	touchRelationship(t, c, "document", "readme", "editor", "user", "alice")

	resp, err := c.CheckPermission(context.Background(), CheckPermissionRequest{
		Resource:   ObjectRef{ObjectType: "document", ObjectID: "readme"},
		Permission: "can_view",
		Subject:    DirectSubject("user", "alice"),
	})
	if err != nil {
		t.Fatalf("CheckPermission: %v", err)
	}
	if !resp.Allowed {
		t.Error("expected computed permission via union to be granted")
	}
}

func containsStr(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > 0 && containsSubstring(s, substr))
}

func containsSubstring(s, sub string) bool {
	for i := 0; i <= len(s)-len(sub); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}
